### PR TITLE
Bug 1937535: retry image pulls during builds

### PR DESCRIPTION
**Pull request by nalind**
Have the builder retry if it fails to pull an image, one which we didn't pre-pull ourselves, during the build.

### CODE REVIEW

**Comment by gabemontero** [comment]
this stems from a conversation with the test platform team around hiccups with pulling in image source content

https://coreos.slack.com/archives/CHY2E1BL4/p1615396283034600 for the tl;dr

@hongkailiu @alvaroaleman FYI

**Comment by gabemontero** [comment]
/assign @adambkaplan 

for approve label

File: pkg/build/builder/source.go
Comment (gabemontero) [inline pkg/build/builder/source.go:451]: ok yep this specifically would have helped the hiccups test-platform is seeing

**Review by gabemontero — APPROVED** [review]
/lgtm

**Comment by gabemontero** [comment]
/title Bug 1937535: retry image pulls during builds

**Comment by gabemontero** [comment]
/retitle Bug 1937535: retry image pulls during builds

**Comment by gabemontero** [comment]
there seems to be some consistent failures in e2e-aws-builds now, but I don't see how these changes could have induced them

**Comment by xiuwang** [comment]
Launching cluster `launch openshift/builder#222` is failed using cluster-bot, I raise issue https://coreos.slack.com/archives/CBN38N3MW/p1615439449195800

**Comment by gabemontero** [comment]
> Launching cluster `launch openshift/builder#222` is failed using cluster-bot, I raise issue https://coreos.slack.com/archives/CBN38N3MW/p1615439449195800

for those who have not opened up that thread, there is a known issue with cluster-bot launching clusters from any PRs

**Comment by gabemontero** [comment]
So I've started debugging the e2e-aws-builds failures.  So far, seems unrelated to this PR.

With the first, `[sig-builds][Feature:Builds][Slow] starting a build using CLI  start-build test context binary builds shoud accept --from-archive with https URL as an input `, there seems to be an unrelated ruby 2.7 hiccup.  

```
2021-03-10T20:17:17.942829557Z STEP 9: RUN scl enable rh-ruby27 "bundle install"
2021-03-10T20:17:19.512648314Z Unable to open /etc/scl/conf/rh-ruby27!
2021-03-10T20:17:26.749584552Z error: build error: error building at STEP "RUN scl enable rh-ruby27 "bundle install"": error while running runtime: exit status 1
```

I looked upstream, and @vrutkovs got a PR merged yesterday, https://github.com/openshift/ruby-hello-world/pull/121, to update things to 2.7.  The way this test works, the archive downloaded is the master branch of this repo.

And the line that fails was one updated by his PR I believe:  https://github.com/openshift/ruby-hello-world/pull/121/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R8

And looking at the buildconfig used for that test, it is still pointing to ruby:2.5 for its from image:  https://github.com/openshift/origin/blob/master/test/extended/testdata/builds/test-build.yaml#L137

So I think e2e-aws-builds will be blocked until we update that test-build.yaml 

Do you already have a PR up for that @vrutkovs ?

`[sig-builds][Feature:Builds][Slow] builds should support proxies start build with broken proxy and a no_proxy override should start a docker build and wait for the build to succeed` is hitting the same thing

and those were the only `sig-builds` failures in that last run

@adambkaplan @bparees FYI

**Comment by vrutkovs** [comment]
Yup, https://github.com/openshift/origin/pull/25949 should move tests to Ruby 2.7

**Comment by gabemontero** [comment]
> Yup, [openshift/origin#25949](https://github.com/openshift/origin/pull/25949) should move tests to Ruby 2.7

ok cool good to know @vrutkovs 

taking a quick peek at that PR, it is big enough that if there is delays in merging it, I might want to pursue a subset PR that fixes the 2 e2e's so we unblock e2e-aws-builds and e2e-gcp-builds ... we'll play it by ear for now

**Review by adambkaplan — APPROVED** [review]
/approve

**Comment by adambkaplan** [comment]
/hold

Looks like hello-openshift updates broke us :/

**Comment by gabemontero** [comment]
the build suite is passing again in https://github.com/openshift/origin/pull/25949

@bparees has tagged that for merge and I'm monitoring it for required vs. non-required e2e semantics

**Comment by gabemontero** [comment]
the test platform guys wanted this backported to 4.7

@hongkailiu @alvaroaleman FYI

/cherrypick release-4.7

**Pull request merged**
